server/shared/src/main/scala/org/http4s/server/middleware/RequestLogger.scala

```diff
@@ -79,35 +79,45 @@ object RequestLogger {
       case Right(_) => true
     }
 
-    Kleisli { req =>
-      if (!logBody) {
-        def logAct = logMessage(req)
-        // This construction will log on Any Error/Cancellation
-        // The Completed Case is Unit, as we rely on the semantics of G
-        // As None Is Successful, but we oly want to log on Some
-        http(req)
-          .guaranteeCase {
-            case Outcome.Succeeded(_) => G.unit
-            case _ => fk(logAct)
-          } <* fk(logAct)
-      } else
-        fk(F.ref(Vector.empty[Chunk[Byte]]))
-          .flatMap { vec =>
-            val collectChunks: Pipe[F, Byte, Nothing] =
-              _.chunks.flatMap(c => Stream.exec(vec.update(_ :+ c)))
-
-            val changedRequest = req.pipeBodyThrough(_.observe(collectChunks))
-
-            val newBody = Stream.eval(vec.get).flatMap(v => Stream.emits(v)).unchunks
-            val logRequest: F[Unit] = logMessage(req.withBodyStream(newBody))
-
-            http(changedRequest)
-              .guaranteeCase {
-                case Outcome.Succeeded(_) => G.unit
-                case _ => fk(logRequest)
+    def logRequest(req: Request[F]): G[Response[F]] = {
+      def logAct = logMessage(req)
+      // This construction will log on Any Error/Cancellation
+      // The Completed Case is Unit, as we rely on the semantics of G
+      // As None Is Successful, but we oly want to log on Some
+      http(req)
+        .guaranteeCase {
+          case Outcome.Succeeded(_) => G.unit
+          case _ => fk(logAct)
+        } <* fk(logAct)
+    }
+
+    Kleisli {
+      case req if !logBody =>
+        logRequest(req)
+      case req =>
+        req.entity match {
+          case Entity.Empty | Entity.Strict(_) =>
+            logRequest(req)
+
+          case Entity.Default(_, _) =>
+            fk(F.ref(Vector.empty[Chunk[Byte]]))
+              .flatMap { vec =>
+                val collectChunks: Pipe[F, Byte, Nothing] =
+                  _.chunks.flatMap(c => Stream.exec(vec.update(_ :+ c)))
+
+                val changedRequest = req.pipeBodyThrough(_.observe(collectChunks))
+
+                val newBody = Stream.eval(vec.get).flatMap(v => Stream.emits(v)).unchunks
+                val logRequest: F[Unit] = logMessage(req.withBodyStream(newBody))
+
+                http(changedRequest)
+                  .guaranteeCase {
+                    case Outcome.Succeeded(_) => G.unit
+                    case _ => fk(logRequest)
+                  }
+                  .map(_.pipeBodyThrough(_.onFinalizeWeak(logRequest)))
               }
```

Comment:
Could we extract the body of this lambda as a function? 
```Scala
def bufferRequest(vv: Vector[Chunk[Byte]]): F[Response[F]]
```

Reply:
Where do you think we should place this function? I'm not too sure about the `Message` as a placeholder. It discloses implementation details of the particular subtype of the `Entity`.

Reply:
Oh, I meant as a nested function of `impl`. Perhaps before the `Kleisli` at line 94, in which case it would also need the request parameter.